### ISSUE TITLE
Emit diagnostic when specializing a non-generic class

### ISSUE THREAD

**Issue by tamireinhorn** [astral-sh/ty#2439]
I wrote a function like:
```py
def fetch_user_countries_count(
    source: str, source_ids: list[int]
) -> sa.Sequence[sa.RowMapping[str, Any]]:
    user_locations = user_author_locations_cte(source=source, source_ids=source_ids)
    stmt = select_countries_with_author_count(user_locations)
    with SESSION() as sess:
        return sess.execute(stmt).mappings().all()
```

And yet, when I call it anywhere else, I get:
```py
def fetch_user_countries_count(
    source: str,
    source_ids: list[int]
) -> @Todo
```

I don't understand what else I should be doing. I tried changing the inner types but to no avail. I saw nothing mentioning said Todo type in the docs as well.

**Comment by AlexWaygood** [comment]
Hey, thanks for the report, and sorry for the confusing UX :-)

The `@Todo` type doesn't mean that you've done anything wrong -- it means that there's some missing logic in ty itself and we don't know how to infer a good type here. To know exactly what the `Todo` is about here, we'll need a way to reproduce this -- how is the variable `sa` defined in your code? Is that a third-party module?

**Comment by tamireinhorn** [comment]
> Hey, thanks for the report, and sorry for the confusing UX :-)
> 
> The `@Todo` type doesn't mean that you've done anything wrong -- it means that there's some missing logic in ty itself and we don't know how to infer a good type here. To know exactly what the `Todo` is about here, we'll need a way to reproduce this -- how is the variable `sa` defined in your code? Is that a third-party module?


yeah, SA is defined as `import sqlalchemy as sa`.

**Comment by dhruvmanila** [comment]
In debug build, it gives me the `@Todo(specialized non-generic class)` type:

```py
import typing as t

import sqlalchemy as sa

def _(x: sa.Sequence[sa.RowMapping[str, t.Any]]):
    # revealed: @Todo(specialized non-generic class)
    reveal_type(x)
```

mypy and Pyright both error on the annotation:
```
mypy: "Sequence" expects no type arguments, but 1 given  [type-arg]
mypy: "RowMapping" expects no type arguments, but 2 given  [type-arg]
Pyright: Expected no type arguments for class "RowMapping" [reportInvalidTypeArguments]
```

ty doesn't raise this error and this suggests that these classes is non-generic, so I think we just need to raise a diagnostic and infer `Unknown` for these cases. In the code, it's marked as TODO:

https://github.com/astral-sh/ruff/blob/09ff3e705639e4773f185994744390fdeb518e8e/crates/ty_python_semantic/src/types/infer/builder/type_expression.rs#L730-L732

https://github.com/astral-sh/ruff/blob/09ff3e705639e4773f185994744390fdeb518e8e/crates/ty_python_semantic/src/types/infer/builder/type_expression.rs#L1066-L1068

I don't see any open issues for this so we can use this.

**Comment by tamireinhorn** [comment]
> In debug build, it gives me the `@Todo(specialized non-generic class)` type:
> 
> ```py
> import typing as t
> 
> import sqlalchemy as sa
> 
> def _(x: sa.Sequence[sa.RowMapping[str, t.Any]]):
>     # revealed: @Todo(specialized non-generic class)
>     reveal_type(x)
> ```
> 
> mypy and Pyright both error on the annotation:
> ```
> mypy: "Sequence" expects no type arguments, but 1 given  [type-arg]
> mypy: "RowMapping" expects no type arguments, but 2 given  [type-arg]
> Pyright: Expected no type arguments for class "RowMapping" [reportInvalidTypeArguments]
> ```
> 
> ty doesn't raise this error and this suggests that these classes is non-generic, so I think we just need to raise a diagnostic and infer `Unknown` for these cases. In the code, it's marked as TODO:
> 
> https://github.com/astral-sh/ruff/blob/09ff3e705639e4773f185994744390fdeb518e8e/crates/ty_python_semantic/src/types/infer/builder/type_expression.rs#L730-L732
> 
> https://github.com/astral-sh/ruff/blob/09ff3e705639e4773f185994744390fdeb518e8e/crates/ty_python_semantic/src/types/infer/builder/type_expression.rs#L1066-L1068
> 
> I don't see any open issues for this so we can use this.

I forgot to mention, but on later testing, I removed the wrong type arguments for RowMapping, and ty still spat out the same type. I put them in just to check what happened and did not remove them from the example posted here. Can edit if preferred.

**Comment by dhruvmanila** [comment]
> I forgot to mention, but on later testing, I removed the wrong type arguments for RowMapping, and ty still spat out the same type. I put them in just to check what happened and did not remove them from the example posted here. Can edit if preferred.

Yeah, I suspect that's because `sa.Sequence` itself is a non-generic class (as highlighted by mypy)?

**Comment by carljm** [comment]
https://github.com/astral-sh/ruff/pull/23767 and https://github.com/astral-sh/ruff/pull/23004 were both abandoned attempts to fix this. Need to look at those more closely to understand why this seems to be tricky. cc @charliermarsh

**Comment by charliermarsh** [comment]
IIRC, it’s because we emit false positives for all the forms we don’t yet support (like TypeVarTuple, and it turned out to be difficult to avoid). I think that at the time, I’d decided it’d be easier to wait for those to be implemented than to workaround them, but I’ll take another look (it’s been a while, and the PR was very stale).

**Comment by carljm** [comment]
Typically the approach to this is to add an explicit match arm for known should-be-but-not-yet-supported forms, that returns `@Todo`, and then switching the fallback case to error. But I assume it turned out to be harder than that for some reason. But I also think it's reasonable to just wait for `TypeVarTuple` support here; that should be coming soon.
